### PR TITLE
[Java] CWE-601: Add Spring URL Redirect ResponseEntity sink

### PR DESCRIPTION
**Pull request by p0wn4j**
Add Open URL Redirect attack sinks by the `ResponseEntity` class in the Spring Framework.

### CODE REVIEW

**Comment by smowton** [comment]
Are you applying to the bounty program with this one?

**Comment by p0wn4j** [comment]
Yes I do.

**Comment by smowton** [comment]
OK, will hold off on reviewing til the security lab have weighed in

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp [hunk 1]
```diff
@@ -22,9 +22,10 @@ redirects on the server; then choose from that list based on the user input prov
 
 <p>The following examples show the bad case and the good case respectively.
 In <code>bad1</code> method and <code>bad2</code> method and <code>bad3</code> method and 
-<code>bad4</code> method, shows an HTTP request parameter being used directly in a URL redirect 
-without validating the input, which facilitates phishing attacks. In <code>good1</code> method, 
-shows how to solve this problem by verifying whether the user input is a known fixed string beginning.
+<code>bad4</code> method and <code>bad5</code> method and <code>bad6</code> method and <code>bad7</code> method,
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp:25]: Change this to `the <code>bad</code> methods`

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll [hunk 2]
```diff
@@ -71,3 +76,66 @@ class SpringUrlRedirectSink extends DataFlow::Node {
...
+ * A sink for URL Redirection via the `ResponseEntity` class.
+ */
+private class SpringResponseEntityUrlRedirectSink extends SpringUrlRedirectSink {
+  SpringResponseEntityUrlRedirectSink() {
+    // Find `new ResponseEntity<>(httpHeaders, ...)` or
+    // `new ResponseEntity<>(..., httpHeaders, ...)` sinks
+    exists(ClassInstanceExpr cie, Argument argument |
+      cie.getConstructedType() instanceof SpringResponseEntity and
+      argument.getType() instanceof SpringHttpHeaders and
+      (argument = cie.getArgument(0) or argument = cie.getArgument(1)) and
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll:90]:
```suggestion
      argument = cie.getArgument([0, 1]) and
```

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll [hunk 3]
```diff
@@ -71,3 +76,66 @@ class SpringUrlRedirectSink extends DataFlow::Node {
...
+      this.asExpr() = argument
+    )
+    or
+    // Find `ResponseEntity.status(...).headers(taintHeaders).build()` or
+    // `ResponseEntity.status(...).location(URI.create(taintURL)).build()` sinks
+    exists(MethodAccess ma, ResponseEntityExpr e |
+      ma = e.getAChildExpr() and
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll:97]: Why are you asking for the `headers` or `location` calls to be a direct child of something of type `SpringResponseEntity`? When is a `headers` or `location` argument not a sink?

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll [hunk 4]
```diff
@@ -71,3 +76,66 @@ class SpringUrlRedirectSink extends DataFlow::Node {
...
+
+private class HttpHeadersSetLocationMethodAccess extends HttpHeadersMethodAccess {
+  HttpHeadersSetLocationMethodAccess() { this.getMethod().hasName("setLocation") }
+}
+
+/**
+ * Holds if `fromNode` to `toNode` is a dataflow step from a tainted argument to
+ * a `HttpHeaders` instance qualifier, i.e. `httpHeaders.setLocation(tainted)`.
+ */
+predicate step(DataFlow::Node fromNode, DataFlow::Node toNode) {
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll:124]: Give this a more informative name, `springUrlRedirectTaintStep`

File: java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp [hunk 5]
```diff
@@ -21,10 +21,10 @@ redirects on the server; then choose from that list based on the user input prov
 <example>
 
 <p>The following examples show the bad case and the good case respectively.
-In <code>bad1</code> method and <code>bad2</code> method and <code>bad3</code> method and 
-<code>bad4</code> method, shows an HTTP request parameter being used directly in a URL redirect 
-without validating the input, which facilitates phishing attacks. In <code>good1</code> method, 
-shows how to solve this problem by verifying whether the user input is a known fixed string beginning.
+In the <code>bad</code> methods it is shown an HTTP request parameter being used directly
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qhelp:24]:
```suggestion
The <code>bad</code> methods show an HTTP request parameter being used directly 
```

**Comment by smowton** [comment]
Test looks to be failing

**Comment by p0wn4j** [comment]
Locally running, test are not failing. Can I see what tests are failing?

**Comment by atorralba** [comment]
> Locally running, test are not failing. Can I see what tests are failing?

The failing test is `ql/java/ql/test/experimental/query-tests/security/CWE-601/SpringUrlRedirect.qlref`.

Seems that a bunch of results are missing:

```
-| SpringUrlRedirect.java:89:38:89:55 | redirectUrl : String | SpringUrlRedirect.java:91:38:91:48 | redirectUrl : String |
-| SpringUrlRedirect.java:91:38:91:48 | redirectUrl : String | SpringUrlRedirect.java:91:27:91:49 | create(...) |
-| SpringUrlRedirect.java:104:39:104:56 | redirectUrl : String | SpringUrlRedirect.java:106:37:106:47 | redirectUrl : String |
-| SpringUrlRedirect.java:106:9:106:19 | httpHeaders [post update] [<map.value>, <element>] : String | SpringUrlRedirect.java:108:68:108:78 | httpHeaders |
-| SpringUrlRedirect.java:106:37:106:47 | redirectUrl : String | SpringUrlRedirect.java:106:9:106:19 | httpHeaders [post update] [<map.value>, <element>] : String |
-| SpringUrlRedirect.java:89:38:89:55 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:91:27:91:49 | create(...) | semmle.label | create(...) |
-| SpringUrlRedirect.java:91:38:91:48 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:104:39:104:56 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:106:9:106:19 | httpHeaders [post update] [<map.value>, <element>] : String | semmle.label | httpHeaders [post update] [<map.value>, <element>] : String |
-| SpringUrlRedirect.java:106:37:106:47 | redirectUrl : String | semmle.label | redirectUrl : String |
-| SpringUrlRedirect.java:108:68:108:78 | httpHeaders | semmle.label | httpHeaders |
-| SpringUrlRedirect.java:91:27:91:49 | create(...) | SpringUrlRedirect.java:89:38:89:55 | redirectUrl : String | SpringUrlRedirect.java:91:27:91:49 | create(...) | Potentially untrusted URL redirection due to $@. | SpringUrlRedirect.java:89:38:89:55 | redirectUrl | user-provided value |
-| SpringUrlRedirect.java:108:68:108:78 | httpHeaders | SpringUrlRedirect.java:104:39:104:56 | redirectUrl : String | SpringUrlRedirect.java:108:68:108:78 | httpHeaders | Potentially untrusted URL redirection due to $@. | SpringUrlRedirect.java:104:39:104:56 | redirectUrl | user-provided value |
```

Maybe you need to rebase/merge `main` to see the failures. You also have some conflicts that need resolution.

**Comment by smowton** [comment]
`java/ql/test/experimental/query-tests/security/CWE-601/SpringUrlRedirect.qlref` still failing. Rebase on latest main?

**Comment by p0wn4j** [comment]
After merge conflict fixing my tests ran ok. Now I  again rebase on lastest main.

**Comment by smowton** [comment]
Reproduced the problem: I was running codeql 2.5.5 and was getting the same results as you; having upgraded to 2.5.7 I now see the failure

**Comment by smowton** [comment]
It was down to a recently-fixed extractor bug that used to misname static inner classes like `Map$Entry` as `Map<>$Entry`. In this case `RequestEntity` no longer needed the superfluous `<>`. I've pushed a commit that fixes the issue.

**Comment by p0wn4j** [comment]
Ok, thank you very much :)

**Review by smowton — APPROVED** [review]

**Pull request merged**